File: selenium-java-demo/src/test/java/pages/DropdownPage.java
package pages;

import io.qameta.allure.Step;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.Select;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Page object for the Dropdown page.
 * <p>
 * Target: <a href="https://the-internet.herokuapp.com/dropdown">/dropdown</a>
 */
public class DropdownPage extends BasePage {

    private static final By DROPDOWN     = By.id("dropdown");
    private static final By PAGE_HEADING = By.tagName("h3");

    public DropdownPage(WebDriver driver) {
        super(driver);
    }

    @Step("Open Dropdown page")
    public DropdownPage open() {
        navigateTo("/dropdown");
        return this;
    }

    @Step("Select dropdown option by value: {value}")
    public DropdownPage selectByValue(String value) {
        Select select = new Select(waitForVisible(DROPDOWN));
        select.selectByValue(value);
        return this;
    }

    @Step("Select dropdown option by visible text: {text}")
    public DropdownPage selectByVisibleText(String text) {
        Select select = new Select(waitForVisible(DROPDOWN));
        select.selectByVisibleText(text);
        return this;
    }

    public String getSelectedOptionText() {
        Select select = new Select(waitForVisible(DROPDOWN));
        return select.getFirstSelectedOption().getText().trim();
    }

    public List<String> getAllOptionTexts() {
        Select select = new Select(waitForVisible(DROPDOWN));
        return select.getOptions().stream()
                .map(WebElement::getText)
                .collect(Collectors.toList());
    }

    public List<String> getSelectableOptionTexts() {
        Select select = new Select(waitForVisible(DROPDOWN));
        return select.getOptions().stream()
                .filter(WebElement::isEnabled)
                .map(WebElement::getText)
                .collect(Collectors.toList());
    }

    public String getPageHeading() {
        return getText(PAGE_HEADING).trim();
    }
}

File: playwright-test-demo/pages/dropdown_page.py
"""
DropdownPage -- Page Object for the Dropdown page.

Target: https://the-internet.herokuapp.com/dropdown
"""

import logging
from typing import List

from pages.base_page import BasePage
from playwright.sync_api import Page

logger = logging.getLogger(__name__)


class DropdownPage(BasePage):
    """Page object for /dropdown."""

    # -- Selectors --------------------------------------------------------- #
    DROPDOWN: str = "#dropdown"
    DROPDOWN_OPTIONS: str = "#dropdown option"
    PAGE_HEADING: str = "h3"

    def __init__(self, page: Page) -> None:
        super().__init__(page)

    # -- Actions ----------------------------------------------------------- #

    def open(self) -> "DropdownPage":
        """Navigate to the dropdown page."""
        self.navigate("/dropdown")
        return self

    def select_option_by_value(self, value: str) -> "DropdownPage":
        """Select a dropdown option by its ``value`` attribute.

        Args:
            value: The ``value`` attribute of the option to select.
        """
        logger.info("Selecting dropdown value='%s'", value)
        self.select_option(self.DROPDOWN, value=value)
        return self

    def select_option_by_visible_text(self, text: str) -> "DropdownPage":
        """Select a dropdown option by its visible text.

        Args:
            text: The visible text of the option to select.
        """
        logger.info("Selecting dropdown label='%s'", text)
        self.select_option(self.DROPDOWN, label=text)
        return self

    # -- State readers ----------------------------------------------------- #

    def get_selected_option_text(self) -> str:
        """Return the visible text of the currently selected option."""
        selected = self.page.locator(f"{self.DROPDOWN} option:checked")
        return selected.inner_text().strip()

    def get_all_option_texts(self) -> List[str]:
        """Return all option texts (including the placeholder)."""
        return self.get_elements_text(self.DROPDOWN_OPTIONS)

    def get_selectable_option_texts(self) -> List[str]:
        """Return option texts excluding disabled placeholder."""
        options = self.page.locator(f"{self.DROPDOWN} option:not([disabled])")
        return options.all_inner_texts()

    def get_page_heading(self) -> str:
        """Return the page heading text."""
        return self.get_text(self.PAGE_HEADING).strip()
